osu.Game/Database/RealmContextFactory.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Threading;
+using osu.Framework.Graphics;
+using osu.Framework.Logging;
+using osu.Framework.Platform;
+using osu.Framework.Statistics;
+using Realms;
+
+namespace osu.Game.Database
+{
+    public class RealmContextFactory : Component, IRealmFactory
+    {
+        private readonly Storage storage;
+
+        private const string database_name = @"client";
+
+        private const int schema_version = 5;
+
+        /// <summary>
+        /// Lock object which is held for the duration of a write operation (via <see cref="GetForWrite"/>).
+        /// </summary>
+        private readonly object writeLock = new object();
+
+        private static readonly GlobalStatistic<int> reads = GlobalStatistics.Get<int>("Realm", "Get (Read)");
+        private static readonly GlobalStatistic<int> writes = GlobalStatistics.Get<int>("Realm", "Get (Write)");
+        private static readonly GlobalStatistic<int> refreshes = GlobalStatistics.Get<int>("Realm", "Dirty Refreshes");
+        private static readonly GlobalStatistic<int> contexts_created = GlobalStatistics.Get<int>("Realm", "Contexts (Created)");
+        private static readonly GlobalStatistic<int> pending_writes = GlobalStatistics.Get<int>("Realm", "Pending writes");
+
+        private Realm context;
```

Comment:
Judging from the test failure this probably needs to get disposed too - albeit weirdly enough I'm not seeing this sort of thing when running tests locally, but since I'm checking on linux it may well be because of different semantics of windows file deletion and unix `unlink()`.

Reply:
having said that, i tried adding a few disposes around the place and it doesn't seem to help - something's apparently still holding files open despite that...

Reply:
Curious, I've looked into this and it seems like the `Dispose`/`Close` flow should be synchronous and quite reliable. I've committed changes to fix the test and the game flow; can you try and see if you can reproduce failures?

Reply:
Seems like it's still occurring. Might need to bring this up as a potential bug in realm, or at least understand the expected behaviour. Will isolate and see if I can reproduce.